### PR TITLE
Add support for scripts, `<script hoist>`

### PR DESCRIPTION
**Pull request by natemoo-re**
I'll walk through this changes in a review!

### CODE REVIEW

File: internal/printer/print-to-js.go [hunk 1]
```diff
@@ -151,31 +140,7 @@ func render1(p *printer, n *Node, opts RenderOptions) {
...
-							p.print(fmt.Sprintf("${%s(\"astro-%s\", ", DEFINE_STYLE_VARS, dataAstroId.Val))
-							p.addSourceMapping(defineVars.ValLoc)
-							p.print(defineVars.Val)
-							p.println(")}")
-						}
-
-						p.addSourceMapping(style.Loc[0])
-						p.print(style.FirstChild.Data)
-						p.addNilSourceMapping()
-						p.println(fmt.Sprintf("%s,", BACKTICK))
```
Comment (natemoo-re) [inline internal/printer/print-to-js.go:178]: Extracted this to shared utility

File: internal/printer/print-to-js.go [hunk 2]
```diff
@@ -237,23 +202,7 @@ func render1(p *printer, n *Node, opts RenderOptions) {
...
-		text := n.Data
-		text = strings.Replace(text, "\\", "\\\\", -1)
-		text = escapeBackticks(text)
-		text = dollarOpen.ReplaceAllString(text, "\\${")
-		text = parenOpen.ReplaceAllString(text, "&lpar;")
-		text = parenClose.ReplaceAllString(text, "&rpar;")
-		text = braceOpen.ReplaceAllString(text, "&lbrace;")
-		text = braceClose.ReplaceAllString(text, "&rbrace;")
-		text = bracketOpen.ReplaceAllString(text, "&lbrack;")
-		text = bracketClose.ReplaceAllString(text, "&rbrack;")
```
Comment (natemoo-re) [inline internal/printer/print-to-js.go:256]: Extracted to an `escapeText` helper

File: internal/printer/printer.go [hunk 3]
```diff
@@ -99,6 +99,63 @@ func (p *printer) printFuncSuffix(componentName string) {
...
+			p.addSourceMapping(a.KeyLoc)
+			p.print(`(` + strings.TrimSpace(a.Key) + `)`)
+		case astro.TemplateLiteralAttribute:
+			p.addSourceMapping(a.KeyLoc)
+			p.print(`"` + strings.TrimSpace(a.Key) + `"`)
+			p.print(":")
+			p.print("`" + strings.TrimSpace(a.Key) + "`")
+		}
+	}
+	p.print("}")
```
Comment (natemoo-re) [inline internal/printer/printer.go:142]: This is a new util as well!

File: internal/printer/printer_test.go [hunk 4]
```diff
@@ -371,7 +372,7 @@ const someProps = {
 
 // Full Astro Component Syntax:
 // https://docs.astro.build/core-concepts/astro-components/`},
-				styles: []string{":root{font-family:system-ui;padding:2em 0;}.counter{display:grid;grid-template-columns:repeat(3,minmax(0,1fr));place-items:center;font-size:2em;margin-top:2em;}.children{display:grid;place-items:center;margin-bottom:2em;}"},
+				styles: []string{fmt.Sprintf(`{props:{"data-astro-id":"HMNNHVCQ"},children:%s:root{font-family:system-ui;padding:2em 0;}.counter{display:grid;grid-template-columns:repeat(3,minmax(0,1fr));place-items:center;font-size:2em;margin-top:2em;}.children{display:grid;place-items:center;margin-bottom:2em;}%s}`, BACKTICK, BACKTICK)},
```
Comment (natemoo-re) [inline internal/printer/printer_test.go:375]: This test shows a big change: we're passing `{props:{},children`code`}` to `$$results.styles` and `$$results.scripts`! This allows us to do hoisting without losing the attributes.

File: internal/printer/printer_test.go [hunk 5]
```diff
@@ -407,6 +408,54 @@ import Widget2 from '../components/Widget2.astro';`},
...
+		{
+			name: "script hoist with frontmatter",
+			source: `---
+---
+<script type="module" hoist>console.log("Hello");</script>`,
+			want: want{
+				imports:     "",
+				frontmatter: []string{},
+				styles:      []string{},
+				scripts:     []string{fmt.Sprintf(`{props:{"type":"module","hoist":true},children:%sconsole.log("Hello");%s}`, BACKTICK, BACKTICK)},
```
Comment (natemoo-re) [inline internal/printer/printer_test.go:420]: As mentioned, we're retaining `type: "module"` and any other attributes. The runtime can remove `hoist` for us.

File: internal/printer/printer_test.go [hunk 6]
```diff
@@ -458,10 +507,17 @@ import Widget2 from '../components/Widget2.astro';`},
...
 				}
 				toMatch = toMatch + STYLE_SUFFIX
 			}
+			if len(tt.want.scripts) > 0 {
+				toMatch = toMatch + SCRIPT_PRELUDE
+				for _, script := range tt.want.scripts {
+					toMatch = toMatch + script + ",\n"
+				}
+				toMatch = toMatch + SCRIPT_SUFFIX
+			}
```
Comment (natemoo-re) [inline internal/printer/printer_test.go:520]: We're actually testing for `tt.want.scripts` now

File: internal/transform/transform.go [hunk 7]
```diff
@@ -22,41 +22,27 @@ func Transform(doc *tycho.Node, opts TransformOptions) {
...
 			switch n.DataAtom {
 			case a.Script:
-				// if <script> has no contents, skip (assume it’s remote)
-				if n.FirstChild == nil {
-					return
+				// if <script hoist>, hoist to the document root
+				if hasTruthyAttr(n, "hoist") {
+					doc.Scripts = append(doc.Scripts, n)
+					// Remove local script node
+					n.Parent.RemoveChild(n)
```
Comment (natemoo-re) [inline internal/transform/transform.go:36]: Only move `<script hoist>` scripts, not other inline scripts

File: internal/transform/utils.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+	for _, attr := range n.Attr {
+		if attr.Key == key &&
+			(attr.Type == astro.EmptyAttribute) ||
+			(attr.Type == astro.ExpressionAttribute && attr.Val == "true") ||
+			(attr.Type == astro.QuotedAttribute && (attr.Val == "" || attr.Val == "true")) {
+			return true
+		}
+	}
+	return false
+}
```
Comment (natemoo-re) [inline internal/transform/utils.go:17]: This is a shared function for checking things like:
```
<style global />
<style global={true} />
<style global="true" />
<style global="" />
<script hoist />
<script hoist={true} />
<script hoist="true" />
<script hoist="" />
```

**Review by natemoo-re — COMMENTED** [review]
Walked through some of the big changes here. Lots of refactoring and bugfixes.

When this is released we'll need to update the REPL and `astro@next` branches to handle the `{props:{}, children: ""}` format for hoisted styles and scripts.

File: internal/printer/printer_test.go (hunk 4 above)
Comment (drwpow) [reply internal/printer/printer_test.go:375]: nit: not crazy about `fmt.Sprintf()` in a test, but it’s fine for this PR. We probably need to split up some of these tests better anyway in the future (not this PR)

File: internal/transform/utils.go (hunk 8 above)
Comment (drwpow) [reply internal/transform/utils.go:17]: That’s pretty cool we retain HTML exactly as-written!

**Review by drwpow — APPROVED** [review]
:shipit:

**Pull request merged**